### ISSUE TITLE
Undefined should be considered equivalent to unassigned

### ISSUE THREAD

**Issue by cefn** [Starcounter-Jack/JSON-Patch#90]
Since an undefined value is not permitted by the JSON specification (only null), I believe that the undefined value stored at a member should be considered equivalent to an unassigned member. Below is a Node REPL which demonstrates the problem and which I've encountered in a production system.

```
> arr
[ , , 'two' ]
> arrUndef
[ undefined, undefined, 'two' ]
> var changes = jsonPatch.compare(arrUndef, arr)
undefined
> changes
[ { op: 'remove', path: '/1' }, { op: 'remove', path: '/0' } ]
```

In practice I've experienced this because of a feature of the LODASH library... 
https://github.com/lodash/lodash/issues/1719
...but if fast-json-patch was stricter in its interpretation of in-memory JSON structures, then this bug/feature would have had significantly less impact on our production system. 

Although no explicit mention is made in the RFC regarding sparse arrays, it should be noted that this behaviour breaks the implied contract that applying a changeset between tree A and tree B to tree A should transform it into tree B, continuing the REPL above...

```javascript
> jsonPatch.apply(arrUndef, changes)
true
> arrUndef
[ 'two' ]
```

**Comment by tomalec** [comment]
Thanks, for feedback @cefn, here is couple of related issues:
https://github.com/Starcounter-Jack/JSON-Patch/issues/60
https://github.com/Starcounter-Jack/JSON-Patch/issues/32

We have plans to support undefined, but it introduced more problems, especially if we would like to keep few constraints:
1. Always produce valid JSON Patches,
2. Keep results of apply and generate consistent. 

For example, if we will say `undefined` <=> unasigned, then for
`A={"smth":"here"}; B=JSON.parse(JSON.stringify(A));` change 
`A.smth = undefined` will produce `remove` operation, then `jsonpatch.apply(B,patch)` will make `B={}`, what led us to inconsistent result.

More results and problems are described in issues mentioned above, and test suites.

**Comment by cefn** [comment]
I guess it's a question of when JSON validity of in-memory structures is imposed, if at all. Currently the reliance on JSON.parse(JSON.stringify(item)) for deepClone makes it impossible to compare in-memory structures with their clones in consistent ways anyway.

For this reason it seemed like a design commitment of fast-json-patch to impose JSON-only for in-memory structures too, meaning that indeed ```A={"smth":undefined}``` is exactly the same data structure as ```A={}``` . If instead your design commitment is the opposite - to support a superset of JSON which preserves in-memory structures (without circular references), then reliance on this clone method in the core has to go.

For reference, in the production system where I'm using fast-json-patch I have monkeypatched your deepClone operation to use npm's require("clone"). If you impose the requirement that patches and clones themselves are JSON, even when held in memory, (implied by the existing deepClone method), then I believe several everyday cases are impossible to solve. If you can afford the dependency on npm clone then progress could be made on it.

The JSON standard seems mute on sparse arrays (absent values, not undefined values), from my minimal exploration, but we definitely need them because of trying to maintain partial trees from a shared JSON tree. However, many implementations, including node's, seem broken when serializing sparse arrays.

We found the below too shocking to be able to rely on the JSON-serialization-clone method for managing or comparing any in-memory structures, with the npm clone implementation being the only one which can correctly recreate the item, and even lodash is shot.

```
 arr = []
[]
> arr[2]="two"
'two'
> clone(arr)
[ , , 'two' ]
> JSON.parse(JSON.stringify(arr))
[ null, null, 'two' ]
> lodash.clone(arr)
[ undefined, undefined, 'two' ]
```

In particular, ```clone(arr)``` creates in-memory duplicates which preserve structure, ( ```clone(arr, false)``` if you want to eliminate JSON-invalid circularity and speed up cloning).

```eval("(" + require("util").inspect(arr) + ")")``` serializes and de-serializes to an equivalent in-memory data structure even for a superset of JSON, which JSON.parse(JSON.stringify(arr)) does not. A minimal implementation of inspect as a standalone module (suitable for browserify) is at npm's util-inspect.
Here's a REPL fragment demonstrating it...

```
> arr = []
> arr[2]="two"
> var inspect = require("util").inspect
> inspect(arr)
'[ , , \'two\' ]'
> eval("(" + inspect(arr) + ")")
[ , , 'two' ]
```

I can understand the intent to use JSON serialization as a reference, but preserving the in-memory structures throughout comparison, patching etc. means that any issues around JSON serialization are encountered only when a user makes a choice of serialisation, not in surprising ways even when these are in-memory structures. 

I believe it's a good idea to support a superset through a more capable clone method as users of the library can always force JSON validity of data or patches by using the JSON-serialize-clone method. Adopting clone means there will be the minimum of surprising bugs from internal memory comparisons, meaning fast-json-patch can support people who want a less-strict superset of JSON too, assuming they manage serialization properly.

**Comment by warpech** [comment]
Fixed in [1.0.0](https://github.com/Starcounter-Jack/JSON-Patch/releases/tag/1.0.0)
